Sources/MapboxDirections/Extensions/Codable.swift

```diff
@@ -0,0 +1,70 @@
+import Foundation
+import Polyline
+import CoreLocation
+
+
+extension Decodable {
+    static internal func from<T: Decodable>(json: String, using encoding: String.Encoding = .utf8) -> T? {
+        guard let data = json.data(using: encoding) else { return nil }
+        return from(data: data) as T?
+    }
+    
+    static internal func from<T: Decodable>(data: Data) -> T? {
+        let decoder = JSONDecoder()
+        return try! decoder.decode(T.self, from: data) as T?
+    }
+}
+
+struct UncertainCodable<T: Codable, U: Codable>: Codable, Equatable {    
+    var t: T?
+    var u: U?
+    
+    var options: DirectionsOptions
+    
+    var value: Codable? {
+        return t ?? u
+    }
+    
+    var coordinates: [CLLocationCoordinate2D] {
+        if let geo = value as? Geometry {
+            return geo.coordinates
+        } else if let geo = value as? String {
+            return decodePolyline(geo, precision: options.shapeFormat == .polyline6 ? 1e6 : 1e5)!
```

Comment:
The current UncertainCodable implementation (and the associated-value enumeration concept we’re bouncing around) would potentially store a Polyline-encoded string and convert it to coordinates on the fly. Given how frequently the navigation SDK gets `coordinates` on a given step, I’m not sure we want to convert on the fly; it could have a significant impact on certain operations like off-route detection.

Reply:
`PolyLineString`, the associated-value enumeration, just stores what the JSON representation does (plus an optional precision). PolyLineString is nothing but an intermediate representation to facilitate conversion from the JSON representation to LineString-backed storage.